TESTS/mbedmicro-rtos-mbed/basic/main.cpp

```diff
@@ -1,38 +1,88 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+/*
+ * Tests is to measure the accuracy of Thread::wait() over a period of time
+ *
+ *
+ * 1) DUT would start to update callback_trigger_count every milli sec
+ * 2) Host would query what is current count base_time, Device responds by the callback_trigger_count
+ * 3) Host after waiting for measurement stretch. It will query for device time again final_time.
+ * 4) Host computes the drift considering base_time, final_time, transport delay and measurement stretch
+ * 5) Finally host send the results back to device pass/fail based on tolerance.
+ * 6) More details on tests can be found in timing_drift_auto.py
+ *
+ */
+
 #include "mbed.h"
 #include "greentea-client/test_env.h"
 #include "rtos.h"
+#include "unity/unity.h"
 
 #if defined(MBED_RTOS_SINGLE_THREAD)
-  #error [NOT_SUPPORTED] test not supported
+#error [NOT_SUPPORTED] test not supported
 #endif
 
-#define TEST_STACK_SIZE 768
+#define TEST_STACK_SIZE 1024
+#define ONE_MILLI_SEC 1000
 
-#define SIGNAL_PRINT_TICK 0x01
+volatile uint32_t callback_trigger_count = 0;
 
-DigitalOut led1(LED1);
+static const int test_timeout = 240;
+bool test_result = false;
 
-const int total_ticks = 10;
-
-void print_tick_thread() {
-    for (int i = 0; i <= total_ticks; i++) {
-      Thread::signal_wait(SIGNAL_PRINT_TICK);
-      greentea_send_kv("tick", i);
-      led1 = !led1;
+void update_tick_thread() {
+    while (true) {
+        Thread::wait(1);
+        ++callback_trigger_count;
     }
 }
 
-int main() {
-    GREENTEA_SETUP(total_ticks + 5, "timing_drift_auto");
+void gt_comm_wait_thread() {
+    char _key[11] = { };
+    char _value[128] = { };
+
+    greentea_send_kv("timing_drift_check_start", 0);
 
-    Thread tick_thread(osPriorityNormal, TEST_STACK_SIZE);
-    tick_thread.start(print_tick_thread);
+    // wait for 1st signal from host
+    greentea_parse_kv(_key, _value, sizeof(_key), sizeof(_value));
+    greentea_send_kv(_key, callback_trigger_count * ONE_MILLI_SEC);
 
-    for (int i = 0; i <= total_ticks; i++) {
-        Thread::wait(1000);
-        tick_thread.signal_set(SIGNAL_PRINT_TICK);
+    // wait for 2nd signal from host
+    greentea_parse_kv(_key, _value, sizeof(_key), sizeof(_value));
+    greentea_send_kv(_key, callback_trigger_count * ONE_MILLI_SEC);
+
+    //get the results from host
+    greentea_parse_kv(_key, _value, sizeof(_key), sizeof(_value));
+
+    if (strcmp("pass", _key) == 0) {
+        test_result = true;
     }
+}
+
+int main() {
+    GREENTEA_SETUP(test_timeout, "timing_drift_auto");
+    Thread tick_thread(osPriorityHigh, TEST_STACK_SIZE);
+    Thread gt_conn_thread(osPriorityNormal, TEST_STACK_SIZE);
+
+    tick_thread.start(update_tick_thread);
+    gt_conn_thread.start(gt_comm_wait_thread);
```

Comment:
I would recommend not creating another thread for this and just to run it on the main stack to reduce memory requirements.